### PR TITLE
Update pygresql from 4.0.0 to 5.1.2

### PR DESCRIPTION
**Pull request by tylarb**
Given that we're already heavily down the path of upgrading pygresql, and given that it has some very consequential changes, I thought it would be good to submit a draft PR of this change.

The most important, cross team changes are as follows:

**Changes to connection object returned by dbconn.connect()**
The changes here are extensive, but most importantly, the connection object no longer closes on `__exit__()` but does commit on `__exit__()`.

In addition, we've added to and cleaned up the helper functions in dbconn.

See the changes and comments in  `gpMgmt/bin/gppylib/db/dbconn.py` file for details 

**Changes to isolation2 testing**

For any team writing tests in isolation2, be aware that the output of pygresql is now quite different from before. 

Previously, `pg.query()` or `cursor.fetchall()` returns an array of strings, each string _very_ roughly equating to a line in `psql` output. Heikki has addressed this previously in [commit 7d7782f1414](https://github.com/greenplum-db/gpdb/commit/7d7782f1414), but it will have to be redone, because the underlying output is different yet again - but hopefully much better in a general sense, because parsing those strings was error prone.

### CODE REVIEW

**Comment by tylarb** [comment]
Pipeline here: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-pygresql2

**Comment by tylarb** [comment]
Before merge, I'll need to update the gitmodules to point back to the master repo for pythonSrc. I'm actually not sure what is happening with the pg_bouncer submodule. It seems that the underlying branch it was pointing to, `pgbouncer_1_8_1`, has been updated, so a submodule update is getting those changes as well.

File: src/test/isolation2/expected/segwalrep/dtx_recovery_wait_lsn.out [hunk 1]
```diff
@@ -148,9 +148,9 @@ DO
 DROP
 4q: ... <quitting>
```
Comment (ashwinstar) [inline src/test/isolation2/expected/segwalrep/dtx_recovery_wait_lsn.out:149]: I don't think this `4q:` is needed. Is it testing anything? I doubt. So, just remove this and it should be good.

Comment (tylarb) [reply src/test/isolation2/expected/segwalrep/dtx_recovery_wait_lsn.out]: Thanks, I've applied this solution. Will need to edit the commit message if it works

File: gpMgmt/bin/analyzedb [hunk 2]
```diff
@@ -167,8 +171,9 @@ def validate_schema_exists(pg_port, dbname, schema):
 def execute_sql(query, master_port, dbname):
     dburl = dbconn.DbURL(port=master_port, dbname=dbname)
     conn = dbconn.connect(dburl)
-    cursor = dbconn.execSQL(conn, query)
-    return cursor.fetchall()
+    result = dbconn.query(conn, query).fetchall()
+    conn.close()
+    return retult
```
Comment (jimmyyih) [inline gpMgmt/bin/analyzedb:176]: Typo here.  Should be `return result`.

File: gpMgmt/bin/gpcheckcat [hunk 3]
```diff
@@ -458,7 +458,7 @@ def getGPConfiguration():
           '''
     curs = db.query(qry)
     for row in curs.dictresult():
-        if row['content'] == -1 and row['isprimary'] != 't':
+        if row['content'] == -1 and not row['isprimary']:
```
Comment (jimmyyih) [inline gpMgmt/bin/gpcheckcat:461]: So results from pygresql are transformed from string to their actual column types now assuming the mapping is available for the column type?

File: gpMgmt/bin/gpcheckcat_modules/unique_index_violation_check.py [hunk 4]
```diff
@@ -39,7 +39,8 @@ def runCheck(self, db_connection):
         violations = []
 
         for (table_oid, index_name, table_name, column_names) in unique_indexes:
-            column_names = column_names[1:-1]
+            delim = ','
+            column_names = delim.join(column_names)
```
Comment (jimmyyih) [inline gpMgmt/bin/gpcheckcat_modules/unique_index_violation_check.py:43]: Could simplify this to `column_names = ",".join(column_names)` which is more Python-like I believe.

File: gpMgmt/bin/gppylib/db/dbconn.py [hunk 5]
```diff
@@ -154,118 +155,160 @@ def canonicalize(s):
...
+        options.append("-c gp_session_role=utility")
 
     # unset search path due to CVE-2018-1058
     if unsetSearchPath:
         options.append("-c search_path=")
 
-    #by default, libpq will print WARNINGS to stdout
-    if not verbose:
-        options.append("-c CLIENT_MIN_MESSAGES=ERROR")
+    # MPP-13779, et al
```
Comment (jimmyyih) [inline gpMgmt/bin/gppylib/db/dbconn.py:226]: Let's get rid of this comment.  It doesn't explain anything even if you go to the JIRA.

File: gpMgmt/bin/gppylib/db/dbconn.py [hunk 6]
```diff
@@ -154,118 +155,160 @@ def canonicalize(s):
...
-    cstr += " options='%s'" % " ".join(options)
+
+    #by default, libpq will print WARNINGS to stdout
+    if not verbose:
+        options.append("-c CLIENT_MIN_MESSAGES=ERROR")
+
+    if options:
+        conninfo['options'] = " ".join(options)
+
+    # MPP-14121, use specified connection timeout
```
Comment (jimmyyih) [inline gpMgmt/bin/gppylib/db/dbconn.py:241]: Let's get rid of the JIRA reference here as well but keep the "Use specified connection timeout" comment.

File: gpMgmt/bin/analyzedb (hunk 2 above)
Comment (tylarb) [reply gpMgmt/bin/analyzedb:176]: Turns out this function isn't used in analyzed - it was imported from `gppylib.operations.backup_utils` at some point, but commit 8922315e34b removed the import from the `backup_utils`. Inside of analyzedb, `execute_sql()` isn't used, probably replaced entirely by `db.query()`

File: gpMgmt/bin/gpcheckcat (hunk 3 above)
Comment (tylarb) [reply gpMgmt/bin/gpcheckcat:461]: That's correct. It's the source of many, many test failures. For more than just column types now - previously, a result might be `'{1,2}'` - that is, a string including curly braces. Now, it is actually `[1,2]`, a list containing ints 1 and 2. I believe only in gpcheckcat was that distinction so onerous that we actually converted back into a string.  All other code should be using native python types. 

See the example where we kept the old formatting here: https://github.com/greenplum-db/gpdb/blob/416bf2fb66bfcace61ac330a4138f7477a50df82/gpMgmt/bin/gpcheckcat#L3392

File: gpMgmt/bin/gpcheckcat_modules/unique_index_violation_check.py (hunk 4 above)
Comment (tylarb) [reply gpMgmt/bin/gpcheckcat_modules/unique_index_violation_check.py:43]: resolved

**Comment by tylarb** [comment]
> Is there any difference between the different connection closing methods seen in this PR?

Yes, absolutely.

First, the `with closing()` syntax requires the closing class from contextlib. As you might expect, it calls `close()` on exiting the with block, which otherwise does not happen. It is functionally identical to the `with conn: ...do work... conn.close()` syntax.

However, the `with conn:` syntax has it's own rule or benefit, which is that at the closing of the `with` block, conn.commit() is called. So you can see you might have multiple `with` blocks before closing the connection if you want to commit more than once.

I also refactored the `execSQL()` function to add the `autocommit=true` value, which means that, outside of a `with` block, `execSQL()` will act like `db.query` from the `pg` module - this means that we can remove the `pg` module entirely, and use only `dbconn`. A use case for this would be transactions like `create database` or `create tablespace` that must be committed immediately - you cannot use a `with conn` to get this behavior, because pygresql dictates that, inside a with block, `conn.commit()` is only called at the exiting of the with block. Thus,

```
conn = dbconn.connect()
dbconn.execSQL(conn, sql)
conn.close()
```
is fundamentally different from a with block - notably, you can create a database or tablespace, which you cannot do in a with block.

You can also override the default behavior of `execSQL()` with the optional `autocommit` arg. We do this in gpexpand - https://github.com/greenplum-db/gpdb/pull/10128/files#diff-8011549bb48f1e35c028e33dd75a053cR1264

In this example, a connection is held open for a long time, but a commit must be explicitly issued.

In the current code, almost all uses of `with dbconn.connect()` could be replaced with a `with closing()` block, but with some refactoring (and better connection management) that could change.

**Comment by tylarb** [comment]
@xiong-gang  commit https://github.com/greenplum-db/gpdb/commit/f3df8b18571dec582dddc60453e8922a8bbbc110 added a test which checks for an exception traceback in `sql_isolation_testcase.py`. I've disabled this test in this PR because the test relies on the full exception stacktrace - which includes file and line numbers, which change any time `sql_isolation_testcase.py` is refactored.

Can you fix the `dtx_recovery_wait_lsn` test so that it does not rely on an exception from the testing framework?

**Pull request merged**